## Lab 3 
### Predicting Digits with Neural Networks

In [2]:
import idx2numpy
import numpy as np

X_train = idx2numpy.convert_from_file('train-images-idx3-ubyte')
y_train = idx2numpy.convert_from_file('train-labels-idx1-ubyte')
X_test = idx2numpy.convert_from_file('t10k-images-idx3-ubyte')
y_test = idx2numpy.convert_from_file('t10k-labels-idx1-ubyte')
X_train = np.reshape(X_train, (60000, 784))#flaten matricies
X_test = np.reshape(X_test,(10000, 784))
print(X_train.shape)
print(y_train.shape)
print(X_test.shape)
print(y_test.shape)

def print_data(y_text, y_predicted):
    print("Accuracy")
    print(accuracy_score(y_test, y_predicted))
    print("Classification")
    print(classification_report(y_test, y_predicted))
    print("Confusion Matrix")
    print(confusion_matrix(y_test, y_predicted))

(60000, 784)
(60000,)
(10000, 784)
(10000,)


In [27]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import classification_report
from sklearn.metrics import confusion_matrix
from sklearn.metrics import accuracy_score
reg = LogisticRegression(solver = 'lbfgs')#code copied from previous lab
reg.fit(X_train, y_train)
y_predicted = reg.predict(X_test)

print_data(y_test, y_predicted)

Accuracy
0.9175
Classification
             precision    recall  f1-score   support

          0       0.95      0.98      0.96       980
          1       0.96      0.98      0.97      1135
          2       0.93      0.88      0.90      1032
          3       0.90      0.91      0.90      1010
          4       0.93      0.93      0.93       982
          5       0.90      0.85      0.88       892
          6       0.94      0.95      0.94       958
          7       0.93      0.92      0.92      1028
          8       0.84      0.88      0.86       974
          9       0.90      0.89      0.90      1009

avg / total       0.92      0.92      0.92     10000

Confusion Matrix
[[ 958    0    0    4    0    3    5    2    6    2]
 [   0 1116    3    1    0    1    4    1    8    1]
 [   8   12  906   18    9    5   10   11   50    3]
 [   3    0   19  916    2   23    5   11   24    7]
 [   1    2    5    3  910    0   11    2   10   38]
 [  11    2    1   40   10  756   16    8   40  

In [28]:
from sklearn.neural_network import MLPClassifier

mlp = MLPClassifier(solver='lbfgs', alpha=1e-5, 
                    hidden_layer_sizes=(50), random_state=1)
mlp.fit(X_train, y_train)
y_predicted = mlp.predict(X_test)

print_data(y_test, y_predicted)

Accuracy
0.8877
Classification
             precision    recall  f1-score   support

          0       0.97      0.93      0.95       980
          1       0.99      0.97      0.98      1135
          2       0.85      0.90      0.87      1032
          3       0.87      0.83      0.85      1010
          4       0.93      0.82      0.87       982
          5       0.86      0.77      0.81       892
          6       0.92      0.94      0.93       958
          7       0.94      0.90      0.92      1028
          8       0.78      0.88      0.83       974
          9       0.80      0.91      0.85      1009

avg / total       0.89      0.89      0.89     10000

Confusion Matrix
[[ 913    0    3    3    0   32   11    2   16    0]
 [   0 1097    8    4    0    3    3    1   18    1]
 [   3    0  926    9   10    0   16    7   55    6]
 [   2    0   70  841    0   33    1   19   35    9]
 [   1    0    3    1  810    3   20    1    6  137]
 [   5    0    6   82    2  686   14    2   77  

In [29]:
mlp = MLPClassifier(solver='lbfgs', alpha=1e-5, 
                    hidden_layer_sizes=(100), random_state=1)
mlp.fit(X_train, y_train)
y_predicted = mlp.predict(X_test)

print_data(y_test, y_predicted)

Accuracy
0.9453
Classification
             precision    recall  f1-score   support

          0       0.97      0.96      0.97       980
          1       0.98      0.98      0.98      1135
          2       0.94      0.95      0.94      1032
          3       0.94      0.93      0.93      1010
          4       0.95      0.94      0.95       982
          5       0.93      0.92      0.93       892
          6       0.96      0.96      0.96       958
          7       0.96      0.93      0.95      1028
          8       0.90      0.94      0.92       974
          9       0.92      0.93      0.92      1009

avg / total       0.95      0.95      0.95     10000

Confusion Matrix
[[ 945    1    5    2    0    3   13    5    6    0]
 [   0 1112    4    3    0    1    2    1   12    0]
 [   4    3  979   11    6    2    4   10   10    3]
 [   1    2   14  942    0   19    2    4   18    8]
 [   0    1   10    0  921    3    6    3    7   31]
 [   5    1    0   26    1  820   12    1   21  

In [30]:
mlp = MLPClassifier(solver='lbfgs', alpha=1e-5, 
                    hidden_layer_sizes=(400), random_state=1)
mlp.fit(X_train, y_train)
y_predicted = mlp.predict(X_test)

print_data(y_test, y_predicted)

Accuracy
0.9678
Classification
             precision    recall  f1-score   support

          0       0.98      0.98      0.98       980
          1       0.99      0.99      0.99      1135
          2       0.97      0.96      0.97      1032
          3       0.96      0.97      0.96      1010
          4       0.97      0.97      0.97       982
          5       0.96      0.95      0.95       892
          6       0.97      0.97      0.97       958
          7       0.97      0.96      0.97      1028
          8       0.96      0.95      0.96       974
          9       0.95      0.96      0.96      1009

avg / total       0.97      0.97      0.97     10000

Confusion Matrix
[[ 965    0    2    2    1    2    3    2    2    1]
 [   0 1120    3    2    0    2    2    3    3    0]
 [   5    3  994    6    4    0    2    7    9    2]
 [   1    0    3  980    0    9    0    3    5    9]
 [   1    0    5    0  952    1    2    4    2   15]
 [   4    0    2   16    2  848    8    1    4  

In [32]:
mlp = MLPClassifier(solver='lbfgs', alpha=1e-5, 
                    hidden_layer_sizes=(100, 50), random_state=1)
mlp.fit(X_train, y_train)
y_predicted = mlp.predict(X_test)

print_data(y_test, y_predicted)

Accuracy
0.9654
Classification
             precision    recall  f1-score   support

          0       0.97      0.98      0.98       980
          1       0.99      0.99      0.99      1135
          2       0.96      0.96      0.96      1032
          3       0.95      0.96      0.96      1010
          4       0.97      0.96      0.96       982
          5       0.95      0.96      0.96       892
          6       0.97      0.97      0.97       958
          7       0.97      0.96      0.97      1028
          8       0.96      0.96      0.96       974
          9       0.95      0.94      0.95      1009

avg / total       0.97      0.97      0.97     10000

Confusion Matrix
[[ 959    0    3    0    2    5    2    6    3    0]
 [   0 1121    5    3    0    1    2    0    2    1]
 [   7    1  989   12    3    2    4    7    7    0]
 [   0    1   14  974    1    6    0    4    8    2]
 [   2    0    5    0  941    0    8    0    2   24]
 [   2    0    1   13    0  859    5    0    6  

### Accuracy

#### Best

With the current trial runs, the 400 layer neural network actually turned out to be the best with an accuracy of 96.78%. However, the two layer neural network was almost just as good with an accuracy of 96.54%

#### Worst

As I would have guessed the worst accuracy was the 50 node neural net. Just looking at that setup it did not seem to have enough nodes to be able to do a decent job at learning the data.

### F1 Score

#### Best 

The best F1 score turned out to be a tie of .97 between the two layer network and the 400 node network. Since the F1 score take precision into account as well, it is actually a better measure of the quality of a model. Thus i find it very interesting that these two different models produced similarly effective models.

#### Worst

As before the 50 node neural network had the worst F1 score. We can conclude from this data that this model is objectivly the worst model we've attempted so far.

### Shape of the Data

The data we used for training this model had 60000 elements with 784 pixels per element. For each of our neural networks we would then have 784 input nodes.

### Why LBFGS

In short we are using LBFGS to optimize our training time. We do this because the training time for these networks can be extremely high to the point that we would spend too much time running the model versus actually studying the data and the result.
LBFGS is an optimization on BFGS. BFGS stores all data points for matrix multiplication whereas LBFGS stores vectors that imply the original data points and is thus able to simplify the calculation and run the training algorithm much quicker.

### Afterthoughts

Neural networks very quickly become a black box. Any typical human being can't take a quick glance at a neural network and say exactly why it came to one conclusion for a peice of data versus another. We can however make some general assumptions about what is happening internally. For instance, THe accuracy, and F1 score of the models increased as the number of nodes increased. With this we can conlude that having too few nodes for a dataset can hinder the performance of a model. There is a limit to this as we saw, I can't just keep adding nodes to a model to increase the accuracy. There was a diminishing return to increasing nodes having a positive result on the outcome. With the limited trial runs it is hard to make a judgement as to why the two layer network worked so comparably to the 400 node model. My best guess would be that the additional layer adds more room for the model to attempt to interpret the data. 
Some potential trials I could run to gain a better grasp on this would be doing a three and four layer run. I would also like to play with having larger first, last, and middle layers to see how that affects the data.


## Tensorflow


In [19]:
#softmax
import tensorflow as tf
from tensorflow.examples.tutorials.mnist import input_data
mnist = input_data.read_data_sets("MNIST_data/", one_hot=True)

x = tf.placeholder(tf.float32, [None, 784])
W = tf.Variable(tf.zeros([784, 10]))
b = tf.Variable(tf.zeros([10]))

Extracting MNIST_data/train-images-idx3-ubyte.gz
Extracting MNIST_data/train-labels-idx1-ubyte.gz
Extracting MNIST_data/t10k-images-idx3-ubyte.gz
Extracting MNIST_data/t10k-labels-idx1-ubyte.gz


In [20]:
#softmax model

y = tf.nn.softmax(tf.matmul(x, W) + b)# create model
y_ = tf.placeholder(tf.float32, [None, 10])
cross_entropy = tf.reduce_mean(-tf.reduce_sum(y_ * tf.log(y), reduction_indices=[1]))
train_step = tf.train.GradientDescentOptimizer(0.5).minimize(cross_entropy)
sess = tf.InteractiveSession()
tf.global_variables_initializer().run()
for _ in range(1000):
  batch_xs, batch_ys = mnist.train.next_batch(100)
  sess.run(train_step, feed_dict={x: batch_xs, y_: batch_ys})
correct_prediction = tf.equal(tf.argmax(y,1), tf.argmax(y_,1))
predicted = tf.argmax(y, 1)
actual = tf.argmax(y_, 1)
print(type(predicted))
pred, act = sess.run(predicted, actual, feed_dict={x: batch_xs, y_: batch_ys})

#print_data(act, pred)
#accuracy = tf.reduce_mean(tf.cast(correct_prediction, tf.float32))
#print(sess.run(accuracy, feed_dict={x: mnist.test.images, y_: mnist.test.labels}))


<class 'tensorflow.python.framework.ops.Tensor'>


TypeError: run() got multiple values for argument 'feed_dict'